src/CachingDocBlockFactory.php

```diff
@@ -0,0 +1,72 @@
+<?php
+declare(strict_types=1);
+
+namespace LanguageServer;
+
+use Microsoft\PhpParser\Node;
+use phpDocumentor\Reflection\DocBlock;
+use phpDocumentor\Reflection\DocBlockFactory;
+use phpDocumentor\Reflection\Types;
+
+/**
+ * Caches DocBlocks by node start position and file URI.
+ */
+class CachingDocBlockFactory
+{
+    /**
+     * Maps file + node start positions to DocBlocks.
+     */
+    private $cache = [];
+
+    /**
+     * @var DocBlockFactory
+     */
+    private $docBlockFactory;
+
+
+    public function __construct()
+    {
+        $this->docBlockFactory = DocBlockFactory::createInstance();
+    }
+
+    /**
+     * @return DocBlock|null
+     */
+    public function getDocBlock(Node $node)
+    {
+        $cacheKey = $node->getStart() . ':' . $node->getUri();
+        if (array_key_exists($cacheKey, $this->cache)) {
```

Comment:
It seems this is a rather hot path, so using
If(isset(...) || array_key_exists(...))
Could speed it up further

Reply:
Interesting - I could try how that affects the performance, but I'm not too optimistic. I think that forming the cache key dominates the performance impact of the function.

Reply:
Is there a reason why only `isset` is not sufficient? `$cache` doesn't have a type declaration unfortunately but I assume it would not contain `null` values?

Reply:
It has `null`s at least for fields where there is no comment text, but the cost for those fields is probably not big, since the comment is never parsed. But there was the following comment in the code, which is the primary reason I added this - I have not verified if that is still the case though:

```php
try {
    // create() throws when it thinks the doc comment has invalid fields.
    // For example, a @see tag that is followed by something that doesn't look like a valid fqsen will throw.
    return $this->docBlockFactory->create($text, $context);
} catch (\InvalidArgumentException $e) {
    return null;
}
```

Reply:
Maybe we could use a NullObject instead of null in this cases?

Reply:
I'm not familiar with the  `NullObject` pattern - so just define `class NullObject {}` and store those in the array, without `NullObject`s ever leaving the cache class? I'm not sure if that would make the code clearer here, since there would have to be a check for `instanceof NullObject` when reading the cache.

Reply:
A NullObject would only use unneeded memory and put pressure on the GC.

Reply:
Not with

```php
class NullObject {
    public static function getInstance(): self {
        static $instance;
        return $instance ?? ($instance = new self);
    }
}
```

However, I'm still not convinced this would make the code clearer, especially since nullness only needs to be checked in one place.